core/src/main/java/org/apache/accumulo/core/dataImpl/KeyExtent.java

```diff
@@ -388,6 +393,70 @@ public Range toMetaRange() {
     return new Range(metadataPrevRow, prevEndRow() == null, toMetaRow(), true);
   }
 
+  /**
+   * Creates a KeyExtent which represents the intersection of this KeyExtent and the passed in
+   * range.
+   * <p>
+   * <b>Note:</b> The range provided must be a range that is derived from a KeyExtent. This means
+   * the range must be in the format of a row range and also requires an exclusive start key, which
+   * is the format that {@link #toDataRange()} uses
+   *
+   * @param range range to clip to
+   * @return the intersection of this KeyExtent and the given range
+   * @throws IllegalArgumentException if the KeyExtent and range do not overlap
+   */
+  public KeyExtent clip(Range range) {
+    return clip(range, false);
+  }
+
+  /**
+   * Creates a KeyExtent which represents the intersection of this KeyExtent and the passed in
+   * range. Unlike {@link #clip(Range)}, this method can optionally return null if the given range
+   * and this KeyExtent do not overlap, instead of throwing an exception. The returnNullIfDisjoint
+   * parameter controls this behavior.
+   * <p>
+   * <b>Note:</b> The range provided must be a range that is derived from a KeyExtent. This means
+   * the range must be in the format of a row range and also requires an exclusive start key, which
+   * is the format that {@link #toDataRange()} uses
+   *
+   * @param range range to clip to
+   * @param returnNullIfDisjoint true to return null if ranges are disjoint, false to throw an
+   *        exception
+   * @return the intersection of this KeyExtent and the given range, or null if given range and this
+   *         KeyExtent do not overlap and returnNullIfDisjoint is true
+   * @throws IllegalArgumentException if the KeyExtent and range does not overlap and
+   *         returnNullIfDisjoint is false
+   *
+   * @see KeyExtent#clip(Range)
+   **/
+  public KeyExtent clip(Range range, boolean returnNullIfDisjoint) {
+    // A standard row range only requires an exclusive key but
+    // KeyExtent has a narrower definition of a row range and toDataRange()
+    // will always create exclusive start keys as well,
+    // so we need to restrict the range further
+    requireRowRange(range);
+    if (!range.isInfiniteStartKey()) {
+      Preconditions.checkArgument(isExclusiveKey(range.getStartKey()),
```

Comment:
If want requireRowRange to be really generic then should also move the isExclusiveKey check for the end key outside to this function.  Looking at the range constructors for rows, it does not always append a 0x00 to the row key for all row ranges.

Reply:
Well, the original intent of requireRowRange was for no-chop merge as there were issues when not requiring that. I don't remember off the top of my head but it was added for a reason to require the end key to be exclusive when we store entries in metadata. So if we remove the exclusive key check then we'd need to add something to still enforce that for TabletFile and StoredTabletFile i think.